Sources/LocalReceiptParsing/Builders/ASN1ContainerBuilder.swift

```diff
@@ -105,16 +105,28 @@ private extension ASN1ContainerBuilder {
             let lengthBytes = data.dropFirst().prefix(totalLengthBytes)
             lengthValue = lengthBytes.toInt()
         }
+
+        var innerContainers: [ASN1Container] = []
         // StoreKitTest receipts report a length of zero for Constructed elements.
         // This is called indefinite-length in ASN1 containers.
-        // When length == 0, the element's contents end when there are two subsequent 0x00 octets.
-        // This (naive) implementation just assumes that the end of content octets are at the end of the
-        // sequence, which is incorrect but works in practice.
-        // This code should be refactored to find the end of content octets, though. 
-        if lengthValue == 0 {
-            lengthValue = data.count - bytesUsedForLength
+        // When length == 0, the element's contents end when there's a container with .endOfContent identifier
+        // To get the length, we build the internal containers until we run into .endOfContent and sum up the bytes used
+        let lengthDefinition: ASN1Length.LengthDefinition = (isConstructed && lengthValue == 0)
+                                                            ? .indefinite : .definite
+
+        if lengthDefinition == .indefinite {
+            innerContainers = try buildInternalContainers(payload: data.dropFirst(bytesUsedForLength))
+            let innerContainersOverallLength = innerContainers.map { $0.totalBytesUsed }.reduce(0, +)
+            lengthValue = innerContainersOverallLength
+        } else if isConstructed {
+            let innerContainerData = data.dropFirst(bytesUsedForLength).prefix(lengthValue)
+            innerContainers = try buildInternalContainers(payload: innerContainerData)
         }
-        return ASN1Length(value: lengthValue, bytesUsedForLength: bytesUsedForLength)
+        let length = ASN1Length(value: lengthValue,
+                                bytesUsedForLength: bytesUsedForLength,
+                                definition: lengthDefinition)
+
+        return (length, innerContainers)
```

Comment:
tests for this coming in as a separate PR so that I can at least have code that I'm not uneasy about in `main`, this code does get the length calculation correctly